packages/aws-cdk/lib/api/util/cloudformation/stack-event-poller.ts

```diff
@@ -88,65 +88,61 @@ export class StackEventPoller {
-        const isParentStackEvent = event.PhysicalResourceId === event.StackId;
-
-        if (isParentStackEvent && this.props.stackStatuses?.includes(event.ResourceStatus ?? '')) {
-          return events;
-        }
-
-        // Fresh event
-        const resEvent: ResourceEvent = {
-          event: event,
-          parentStackLogicalIds: this.props.parentStackLogicalIds ?? [],
-          isStackEvent: isParentStackEvent,
-        };
-        events.push(resEvent);
-
-        if (
-          !isParentStackEvent &&
-          event.ResourceType === 'AWS::CloudFormation::Stack' &&
-          isStackBeginOperationState(event.ResourceStatus)
-        ) {
-          // If the event is not for `this` stack and has a physical resource Id, recursively call for events in the nested stack
-          this.trackNestedStack(event, [...(this.props.parentStackLogicalIds ?? []), event.LogicalResourceId ?? '']);
-        }
-
-        if (isParentStackEvent && isStackTerminalState(event.ResourceStatus)) {
-          this.complete = true;
+      for await (const page of paginator) {
+        for (const event of page?.StackEvents ?? []) {
+          // Event from before we were interested in 'em
+          if (this.props.startTime !== undefined && event.Timestamp!.valueOf() < this.props.startTime) {
+            return events;
+          }
+
+          // Already seen this one
+          if (this.eventIds.has(event.EventId!)) {
+            return events;
+          }
+          this.eventIds.add(event.EventId!);
+
+          // The events for the stack itself are also included next to events about resources; we can test for them in this way.
+          const isParentStackEvent = event.PhysicalResourceId === event.StackId;
+
+          /* istanbul ignore next */
+          if (isParentStackEvent && this.props.stackStatuses?.includes(event.ResourceStatus ?? '')) {
+            return events;
+          }
+
+          // Fresh event
+          const resEvent: ResourceEvent = {
+            event: event,
+            parentStackLogicalIds: this.props.parentStackLogicalIds ?? [],
+            isStackEvent: isParentStackEvent,
+          };
+          events.push(resEvent);
+
+          if (
+            !isParentStackEvent &&
+            event.ResourceType === 'AWS::CloudFormation::Stack' &&
+            isStackBeginOperationState(event.ResourceStatus)
+          ) {
+            /* istanbul ignore next */
+            // If the event is not for `this` stack and has a physical resource Id, recursively call for events in the nested stack
+            this.trackNestedStack(event, [...(this.props.parentStackLogicalIds ?? []), event.LogicalResourceId ?? '']);
+          }
+
+          /* istanbul ignore next */
+          if (isParentStackEvent && isStackTerminalState(event.ResourceStatus)) {
+            this.complete = true;
+          }
         }
       }
     } catch (e: any) {
+      /* istanbul ignore next */
```

Comment:
Already [uncovered](https://app.codecov.io/github/aws/aws-cdk/blob/main/packages%2Faws-cdk%2Flib%2Fapi%2Futil%2Fcloudformation%2Fstack-event-poller.ts#L135) on `main` but codecov thinks its a new line because of the refactor.